plugin/src/compilation/decompiler.ml

```diff
@@ -0,0 +1,89 @@
+
+open Names
+open Constr
+open Environ
+open Envutils
+open Tactics
+open Pp
+open Contextutils
+
+(* Abstraction of Coq tactics supported by this decompiler. *)
+type tact =
+  | Intro of Id.t
+  | Intros of Id.t list
+  | Apply of env * types
+        
+(* Return the string representation of a single tactic.
+   TODO: Indentation/bulleting
+         Don    't expose this in .mli *)
+let show_tact sigma tac : Pp.t =
+  (match tac with
+   | Intro n -> str ("intro " ^ Id.to_string n)
+   | Intros ns ->
+      let names = String.concat " " (List.map Id.to_string ns) in
+      str ("intros " ^ names)
+   | Apply (env, trm) ->
+      let body_s = Printer.pr_constr_env env sigma trm in
+      str "apply " ++ body_s)
+  ++ str ".\n" (* maybe ";" in future *)
+
+(* Converts "intro n. intro m. ..." into "intros n m ..." *)
+let rec collapse_intros (tacs : tact list) : tact list =
+  (*match tacs with
```

Comment:
Ah, I see. So you build it as a list of "intro" tactics and then postprocess. I could see that being easier than passing around a list.